File: 프로그래머스/lv2/42842. 카펫/카펫.js
function solution(brown, yellow) {
    const dividers = getDividers(yellow);
    const row_col = (brown - 4)/2;
    
    const result = dividers.filter(([row, col]) => row + col === row_col);
    [row, col] = result[0];
    return [row + 2, col + 2];
}

const getDividers = num => {
    const dividers = [];
    for(let i = 1; i <= Math.sqrt(num); i++) {
        if(num % i === 0) {
            dividers.push([num / i , i])
        }
    }
    return dividers;
}
/*
    노랑의 약수들을 구한다.
    모서리를 제외하기 위해 갈색 - 4
    가로 + 세로를 구하기 위해 갈색 / 2
    
    앞서 구한 노랑 약수 짝들이 더해서 갈색(가로 + 세로)가 되는 값을 구한다.
    이 값 + 2씩 하면 카펫 가로, 세로
*/

File: 프로그래머스/2/42842. 카펫/카펫.py
'''
반을 나눌 수 있다.
(r - 2)(c - 2) = y
2c + 2r - 4 = b

근의 공식을 이용하면 된다. 
그러니 a, b, c를 먼저 구하자. 

[a]
(b + 4) / 2 = c + r
r = ((b + 4) / 2) - c

rc - 2c - 2r + 4 = y
[a = 1]

[b]
c(((b + 4) / 2) - c) - 2c - 2(((b + 4) / 2) - c) + 4 = y
- c^2 + ((b + 4) / 2)c -2((b + 4) / 2) + 4 = y
[b = -((b + 4) / 2)]

[c]
[c = y - 4 + 2((b + 4) / 2)]
'''
import math

def solution(brown, yellow):
    a = 1
    b = -((brown + 4) / 2)
    c = yellow - 4 + 2 * ((brown + 4) / 2)
    
    r1 = (-b + math.sqrt(b**2 - 4 * a * c)) / 2 * a
    r2 = (-b - math.sqrt(b**2 - 4 * a * c)) / 2 * a
    
    if r1 > r2:
        return [r1, r2]
    else:
        return [r2, r1]
    
    
    